_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+
+## Level 3 - Hypermedia Controls
+
+Hypermedia systems extend the resource state that is being manipulated by the services with additional characteristics. 
+The resource state becomes a combination of:
+* The values of the individual variables that make up the resource
+* Links to related resources
+* Links to manipulate the current resource (creating, updating, deleting…)
+* Evaluation results of business rules encompassing the resource and other related resources
+
+The links mentioned make up a domain application protocol (DAP) that advertises all possible interactions with the resource. 
+The consumers of this service use these published interactions to discover how to interact with the resource. 
+Hyperlinks are the weapon of choice when adding such links to interactions. 
+A DAP also consists of two other key components: Media Types (as mentioned before), and HTTP Idioms (which make up the HTTP uniform interface: verbs, standard headers, error codes…). 
+In short: Link relation values tell why the consumer should activate a hypermedia control by stating the role of the linked resource, the media type tells the consumer what it can expect as the response of a link, and the idioms manipulate the resources represented by the links.
+
+The resource state can be in any format as REST’s hypermedia tenet does not force any set media type. 
+IANA is the official registry of MIME media types and maintains a [list of all the universally accepted MIME types](http://www.iana.org/assignments/media-types/media-types.xhtml){:target="_blank" rel="noopener noreferrer"}.
+The resource should indicate in the Content-Type header of its responses the ideal way of interpreting it. 
+Similarly, the Accept header is part of the HTTP spec for requests. 
+In this header the consumer can indicate which media type it wishes to receive. 
+Based on these two headers, a system of content negotiation is set up where the request of a consumer can be matched to the proper implementation to return the desired result and how it is conveyed.
+
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2021-01-04-Designing-REST-services/mediatypes.png' | prepend: site.baseurl }}" alt="Workspace" class="image" style="margin:0px auto; max-width:100%">
+{: refdef}
+
+The HTTP specification allows for the definition of custom media types as well. 
+This is used to inform the consumer about the result to be expected in greater detail. 
+Where “application/xml” informs the consumer which format to expect, it doesn’t give any insight on which data to expect. 
+This can be done by creating a custom media type in the vendor range (vnd). 
+This type adheres to the format:
+
+```
+<media type>/vnd.<owner of custom media type>.<type of data>+<media suffix>
+```
+
+For example: If the return of a service would list the detail of a consultant for the company “Best IT People” in XML format, you would get the following media type: “application/vnd.bestitpeople.consultant+xml”. 
```

Comment:
I would also use backticks here for better readability.
```suggestion
For example: If the return of a service would list the detail of a consultant for the company “Best IT People” in XML format, you would get the following media type: `application/vnd.bestitpeople.consultant+xml`. 
```

Reply:
done